File: src/AccessManager.Api/Responses/ApiResponse.cs
namespace AccessManager.Api.Responses;

public class ApiResponse
{
    public bool Success { get; set; }
    public IReadOnlyCollection<string> Errors { get; set; } = [];

    public static ApiResponse Ok()
    {
        return new ApiResponse
        {
            Success = true
        };
    }

    public static ApiResponse Fail(IReadOnlyCollection<string> errors)
    {
        return new ApiResponse
        {
            Success = false,
            Errors = errors
        };
    }
}

public class ApiResponse<T>
{
    public bool Success { get; set; }
    public T? Data { get; set; }
    public IReadOnlyCollection<string> Errors { get; set; } = [];

    public static ApiResponse<T> Ok(T data)
    {
        return new ApiResponse<T>
        {
            Success = true,
            Data = data
        };
    }

    public static ApiResponse<T> Fail(IReadOnlyCollection<string> errors)
    {
        return new ApiResponse<T>
        {
            Success = false,
            Errors = errors
        };
    }
}

File: frontend/src/api/apiResponse.js
export function getApiErrors(responseData) {
  if (Array.isArray(responseData?.errors) && responseData.errors.length > 0) {
    return responseData.errors.join(' ')
  }

  return 'Erro ao concluir a operacao.'
}

export function unwrapApiResponse(response, fallbackData) {
  if (response.data?.success === false) {
    throw new Error(getApiErrors(response.data))
  }

  return response.data?.data ?? fallbackData
}

export function handleApiError(error) {
  if (error.response?.data) {
    throw new Error(getApiErrors(error.response.data))
  }

  throw new Error('Erro ao conectar com a API.')
}
